### ISSUE TITLE
let getAsync and putAsync return ICompletableFuture

### ISSUE THREAD

**Issue by valdo404** [hazelcast/hazelcast#5315]
As for now it is very complicated to listen a getAsync or putAsync result and to integrate it with completable futures or listenable futures.

Would it be possible to return an ICompletableFuture since it is an interface which seems to extends jdk futures and is returned by an IMap ?

**Comment by gurbuzali** [comment]
you can cast to `ICompletableFuture` as a workaround

**Comment by jerrinot** [comment]
Hi @valdo404,
I understand the need, however it's quite tricky - returning a different type breaks binary compatibility. It doesn't matter the new type extents the original type - they are treated as different types when it comes to a method dispatch. 

As @gurbuzali wrote you can cast return values of most async methods to ICompletableFuture - just be aware that doing so may slightly increase a risk of a migration to a newer version - as it relies on implementation details.

**Comment by valdo404** [comment]
Well, considering futures and asynchronous requests, do you have a plan for the whole thing ? Sometimes when the data is loaded or written to/from a map loader, synchronous requests are bad.

**Comment by Tembrel** [comment]
Casting with a fallback should be safe enough:

````java
    public static <T> ICompletableFuture<T> toCompletableFuture(
            Future<T> f, ExecutorService e) {

        if (f instanceof ICompletableFuture) {
            return (ICompletableFuture<T>) f;
        } else {
            return new CompletableFutureTask<>(() -> f.get(), e);
        }
    }
    
    public static <T> ICompletableFuture<T> toCompletableFuture(Future<T> f) {
        return toCompletableFuture(f, ForkJoinPool.commonPool());
    }
````

Note that I'm using lambda syntax from Java 8 for compactness, but this approach really only makes sense for Java 6/7 uses. (And the common pool is only available in Java 8, so you'd have to find your own default here.)

`ICompletableFuture` is an awkward but unfortunately necessary wart in migrating towards full use of Java 8 while retaining Java 6/7 compatibility. If you're using Java 8 and Hazelcast (as I am), you'll want to avoid it and move to `java.util.concurrent.CompletableFuture` wherever possible, hiding the details behind adapter functions analogous to the one above but using true Java 8 abstractions.

**Comment by LoneRifle** [comment]
This issue is a duplicate of - and superseded by - #6999, and will be implemented by #7967.

**Comment by vbekiaris** [comment]
Fixed by #7967
